<?xml version="1.0"?>
<robot name="lebai_lm3_l1_with_doser_and_camera">

  <!-- Materials -->
  <material name="white"><color rgba="0.86 0.85 0.81 1.0"/></material>
  <material name="light_grey"><color rgba="0.61 0.66 0.67 1"/></material>
  <material name="light_blue"><color rgba="0.79 0.82 0.93 1"/></material>
  <material name="metal_grey"><color rgba="0.60 0.67 0.69 1"/></material>

  <!-- World -->
  <link name="world"/>

  <!-- Base Link -->
  <link name="base_link">
    <inertial>
      <origin xyz="0 0 0.05" rpy="0 0 0"/>
      <mass value="2.0"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/visual/base_link.stl"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/collision/base_link.stl"/></geometry>
    </collision>
  </link>

  <link name="link_1">
    <inertial>
      <origin xyz="0.0 -0.004834 -0.08972" rpy="0 0 0"/>
      <mass value="2.31"/>
      <inertia ixx="0.0093" ixy="0" ixz="0" iyy="0.00913" iyz="0" izz="0.00303"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/visual/link1.stl"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/collision/link1.stl"/></geometry>
    </collision>
  </link>

  <link name="link_2">
    <inertial>
      <origin xyz="-0.18435 0.0 0.09428" rpy="0 0 0"/>
      <mass value="2.1612"/>
      <inertia ixx="0.0211" ixy="0" ixz="0.03767" iyy="0.1554" iyz="0" izz="0.13643"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/visual/link2.stl"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/collision/link2.stl"/></geometry>
    </collision>
  </link>

  <link name="link_3">
    <inertial>
      <origin xyz="-0.14386 0.0 0.01892" rpy="0 0 0"/>
      <mass value="1.8508"/>
      <inertia ixx="0.002519" ixy="0" ixz="0.00121" iyy="0.08222" iyz="0" izz="0.08138"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/visual/link3.stl"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/collision/link3.stl"/></geometry>
    </collision>
  </link>

  <link name="link_4">
    <inertial>
      <origin xyz="0.0 0.011 -0.03772" rpy="0 0 0"/>
      <mass value="0.93256"/>
      <inertia ixx="0.0032645" ixy="0" ixz="0" iyy="0.003061" iyz="0" izz="0.000866"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/visual/link4.stl"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/collision/link4.stl"/></geometry>
    </collision>
  </link>

  <link name="link_5">
    <inertial>
      <origin xyz="0.0 -0.011 -0.03772" rpy="0 0 0"/>
      <mass value="0.93256"/>
      <inertia ixx="0.0032645" ixy="0" ixz="0" iyy="0.003061" iyz="0" izz="0.000866"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/visual/link5.stl"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/collision/link5.stl"/></geometry>
    </collision>
  </link>

  <link name="link_6">
    <inertial>
      <origin xyz="0 0 -0.04882" rpy="0 0 0"/>
      <mass value="0.5782"/>
      <inertia ixx="0.001883" ixy="0" ixz="0" iyy="0.0018889" iyz="0" izz="0.000404"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/visual/link6.stl"/></geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/collision/link6.stl"/></geometry>
    </collision>
  </link>

  <link name="base"><inertial><mass value="0.001"/><inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/></inertial></link>
  <link name="tool0"><inertial><mass value="0.001"/><inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/></inertial></link>

  <!-- Joints -->
  <joint name="world_joint" type="fixed">
    <parent link="world"/><child link="base_link"/>
    <origin xyz="0 0 0.01" rpy="0 0 0"/>
  </joint>

  <joint name="joint_1" type="revolute">
    <parent link="base_link"/><child link="link_1"/>
    <origin xyz="0 0 0.21583" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-6.28" upper="6.28" effort="100" velocity="3.14"/>
    <dynamics damping="0.5" friction="0.1"/>
  </joint>

  <joint name="joint_2" type="revolute">
    <parent link="link_1"/><child link="link_2"/>
    <origin xyz="0 0 0" rpy="1.570796327 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-6.28" upper="6.28" effort="100" velocity="3.14"/>
    <dynamics damping="0.5" friction="0.1"/>
  </joint>

  <joint name="joint_3" type="revolute">
    <parent link="link_2"/><child link="link_3"/>
    <origin xyz="-0.38 0 0" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-6.28" upper="6.28" effort="100" velocity="3.14"/>
    <dynamics damping="0.5" friction="0.1"/>
  </joint>

  <joint name="joint_4" type="revolute">
    <parent link="link_3"/><child link="link_4"/>
    <origin xyz="-0.355 0 0.12063" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-6.28" upper="6.28" effort="100" velocity="3.14"/>
    <dynamics damping="0.5" friction="0.1"/>
  </joint>

  <joint name="joint_5" type="revolute">
    <parent link="link_4"/><child link="link_5"/>
    <origin xyz="0 -0.09833 0" rpy="1.570796327 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-6.28" upper="6.28" effort="5.5" velocity="3.14"/>
    <dynamics damping="0.5" friction="0.1"/>
  </joint>

  <joint name="joint_6" type="revolute">
    <parent link="link_5"/><child link="link_6"/>
    <origin xyz="0 0.08343 0" rpy="-1.570796327 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-6.28" upper="6.28" effort="4.6" velocity="3.14"/>
    <dynamics damping="0.5" friction="0.1"/>
  </joint>

  <joint name="joint_6-tool0" type="fixed">
    <parent link="link_6"/><child link="tool0"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <joint name="base_link-base" type="fixed">
    <parent link="base_link"/><child link="base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <!-- Gripper -->
  <link name="g_base_link">
    <inertial>
      <origin xyz="0.0006 -0.0076 0.0407" rpy="0 0 0"/>
      <mass value="1.01"/>
      <inertia ixx="0.000786" ixy="-5.4e-6" ixz="-3e-6" iyy="0.000513" iyz="-0.000287" izz="0.000431"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0.0007" rpy="0 0 0"/>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/Gripper_link.STL"/></geometry>
      <material name="light_blue"/>
    </visual>
    <collision>
      <origin xyz="0 0 0.0007" rpy="0 0 0"/>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/Gripper_link.STL"/></geometry>
    </collision>
  </link>

  <link name="g_brush_link">
    <inertial>
      <origin xyz="0 0 0.0017" rpy="0 0 0"/>
      <mass value="0.123"/>
      <inertia ixx="4.95e-5" ixy="0" ixz="0" iyy="4.95e-5" iyz="0" izz="4.45e-6"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0.0017" rpy="0 0 0"/>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/Brush_link.STL"/></geometry>
      <material name="light_blue"/>
    </visual>
    <collision>
      <origin xyz="0 0 0.0017" rpy="0 0 0"/>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/Brush_link.STL"/></geometry>
    </collision>
  </link>

  <link name="g_end_link">
    <inertial>
      <origin xyz="0.0048 0.0122 -0.1385" rpy="0 0 0"/>
      <mass value="1.14"/>
      <inertia ixx="0.00274" ixy="-0.0001" ixz="0.0001" iyy="0.00248" iyz="0.000327" izz="0.000518"/>
    </inertial>
  </link>

  <joint name="gripper_mount_joint" type="fixed">
    <parent link="tool0"/><child link="g_base_link"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <joint name="g_brush_joint" type="fixed">
    <parent link="g_base_link"/><child link="g_brush_link"/>
    <origin xyz="0 0.033 0.1026" rpy="0.756 -0.236 -2.899"/>
  </joint>

  <joint name="g_end_link_joint" type="fixed">
    <parent link="g_brush_link"/><child link="g_end_link"/>
    <origin xyz="0 0 0.0989" rpy="0 0 3.14159"/>
  </joint>

  <!-- Camera System -->
  <link name="cam_case_base_link">
    <inertial>
      <origin xyz="-0.0003 -0.385 -0.001" rpy="0 0 0"/>
      <mass value="0.996"/>
      <inertia ixx="0.0529" ixy="0.0006" ixz="0" iyy="0.0529" iyz="0" izz="0.1059"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/case_base_link.STL"/></geometry>
      <material name="light_grey"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/case_base_link.STL"/></geometry>
    </collision>
  </link>

  <link name="cam_kamera_dikey_aparat">
    <inertial>
      <origin xyz="0 0 0.5" rpy="0 0 0"/>
      <mass value="0.4"/>
      <inertia ixx="0.0333" ixy="0" ixz="0" iyy="0.0333" iyz="0" izz="2.67e-5"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/kamera_dikey_aparat.STL"/></geometry>
      <material name="metal_grey"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/kamera_dikey_aparat.STL"/></geometry>
    </collision>
  </link>

  <link name="cam_kamera_yatay_aparat">
    <inertial>
      <origin xyz="0.0001 0.15 -0.0087" rpy="0 0 0"/>
      <mass value="0.297"/>
      <inertia ixx="0.00223" ixy="0" ixz="0" iyy="3.08e-5" iyz="0" izz="0.00225"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/kamera_yatay_aparat.STL"/></geometry>
      <material name="metal_grey"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/kamera_yatay_aparat.STL"/></geometry>
    </collision>
  </link>

  <link name="cam_kamera_tutucu">
    <inertial>
      <origin xyz="0 0 0.004" rpy="0 0 0"/>
      <mass value="0.0032"/>
      <inertia ixx="1.24e-7" ixy="0" ixz="0" iyy="1.24e-7" iyz="0" izz="2.13e-7"/>
    </inertial>
    <visual>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/kamera_tutucu.STL"/></geometry>
      <material name="light_blue"/>
    </visual>
    <collision>
      <geometry><mesh filename="file:///home/can/Aware_ws/src/lebai_lm3_support/meshes/lm3_l1/kamera_tutucu.STL"/></geometry>
    </collision>
  </link>

  <link name="cam_kamera_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="white"/>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <!-- camera -->
  <gazebo reference="cam_kamera_link">
    <sensor type="camera" name="camera1">
      <update_rate>30.0</update_rate>
      <camera name="head">
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>800</width>
          <height>800</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>300</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <!-- Noise is sampled independently per pixel on each frame.
               That pixel's noise value is added to each of its color
               channels, which at that point lie in the range [0,1]. -->
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>rrbot/camera1</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>camera_link</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>

  <joint name="cam_case_base_joint" type="fixed">
    <parent link="base"/><child link="cam_case_base_link"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <joint name="cam_kamera_dikey_aparat_joint" type="fixed">
    <parent link="cam_case_base_link"/><child link="cam_kamera_dikey_aparat"/>
    <origin xyz="-0.24788 -0.63459 0" rpy="0 0 -0.7854"/>
  </joint>

  <joint name="cam_kamera_yatay_aparat_joint" type="fixed">
    <parent link="cam_kamera_dikey_aparat"/><child link="cam_kamera_yatay_aparat"/>
    <origin xyz="0 0.01 0.735" rpy="0 0 0"/>
  </joint>

  <joint name="cam_kamera_tutucu_joint" type="fixed">
    <parent link="cam_kamera_yatay_aparat"/><child link="cam_kamera_tutucu"/>
    <origin xyz="0 0.235 -0.01" rpy="3.1416 0 0"/>
  </joint>

  <joint name="cam_kamera_link_joint" type="fixed">
    <parent link="cam_kamera_tutucu"/><child link="cam_kamera_link"/>
    <origin xyz="0 0 0.025" rpy="1.5707 4.71219 0"/>
  </joint>

  <!-- Gazebo Colors -->
  <gazebo reference="base_link"><material>Gazebo/White</material></gazebo>
  <gazebo reference="link_1"><material>Gazebo/White</material></gazebo>
  <gazebo reference="link_2"><material>Gazebo/White</material></gazebo>
  <gazebo reference="link_3"><material>Gazebo/White</material></gazebo>
  <gazebo reference="link_4"><material>Gazebo/White</material></gazebo>
  <gazebo reference="link_5"><material>Gazebo/White</material></gazebo>
  <gazebo reference="link_6"><material>Gazebo/White</material></gazebo>
  <gazebo reference="g_base_link"><material>Gazebo/SkyBlue</material></gazebo>
  <gazebo reference="g_brush_link"><material>Gazebo/SkyBlue</material></gazebo>
  <gazebo reference="cam_case_base_link"><material>Gazebo/Grey</material></gazebo>
  <gazebo reference="cam_kamera_dikey_aparat"><material>Gazebo/Grey</material></gazebo>
  <gazebo reference="cam_kamera_yatay_aparat"><material>Gazebo/Grey</material></gazebo>
  <gazebo reference="cam_kamera_tutucu"><material>Gazebo/SkyBlue</material></gazebo>
  <gazebo reference="cam_kamera_link"><material>Gazebo/SkyBlue</material></gazebo>

  <!-- ============================================ -->
  <!-- Gazebo ros2_control Plugin                   -->
  <!-- ============================================ -->
  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>/home/can/Aware_ws/src/lebai_lm3_gazebo/config/controllers.yaml</parameters>
    </plugin>
  </gazebo>

  <!-- ============================================ -->
  <!-- ros2_control Hardware Interface              -->
  <!-- ============================================ -->
  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>

    <joint name="joint_1">
      <command_interface name="position">
        <param name="min">-6.28</param>
        <param name="max">6.28</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">0.2</param> </state_interface>
      <state_interface name="velocity"/>
    </joint>

    <joint name="joint_2">
      <command_interface name="position">
        <param name="min">-6.28</param>
        <param name="max">6.28</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">-1.42</param> </state_interface>
      <state_interface name="velocity"/>
    </joint>

    <joint name="joint_3">
      <command_interface name="position">
        <param name="min">-6.28</param>
        <param name="max">6.28</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">1.85</param> </state_interface>
      <state_interface name="velocity"/>
    </joint>

    <joint name="joint_4">
      <command_interface name="position">
        <param name="min">-6.28</param>
        <param name="max">6.28</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">-0.42</param> </state_interface>
      <state_interface name="velocity"/>
    </joint>

    <joint name="joint_5">
      <command_interface name="position">
        <param name="min">-6.28</param>
        <param name="max">6.28</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">0.2</param> </state_interface>
      <state_interface name="velocity"/>
    </joint>

    <joint name="joint_6">
      <command_interface name="position">
        <param name="min">-6.28</param>
        <param name="max">6.28</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">0.0</param> </state_interface>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>

</robot>
